chain/contracts/ScorchedEarth.sol

```diff
@@ -9,14 +9,182 @@ import "@openzeppelin/contracts-ethereum-package/contracts/math/SafeMath.sol";
 contract ScorchedEarth is ForceMoveApp {
     using SafeMath for uint256;
 
+    enum Phase {
+        Suggest,
+        React
+    }
+
+    enum Reaction {
+        None, // Not applicable in Suggest phase
+        Reward,
+        Punish
+    }
+
+    struct SEData {
+        uint256 payment;
+        uint256 suggesterBurn;
+        uint256 userBurn;
+        Phase phase;
+        Reaction reaction;
+        string suggestion;
+    }
+
     constructor () public { }
 
+    function appData(bytes memory appDataBytes) internal pure returns (SEData memory) {
+        return abi.decode(appDataBytes, (SEData));
+    }
+
     function validTransition(
         VariablePart memory _fromPart,
         VariablePart memory _toPart,
         uint48, // turnNumB
         uint256  // nParticipants
     ) public pure override returns (bool) {
+        Outcome.AllocationItem[] memory fromAllocation = extractAllocation(_fromPart);
+        Outcome.AllocationItem[] memory toAllocation = extractAllocation(_toPart);
+
+        requireDestinationsUnchanged(fromAllocation, toAllocation);
+
+        // decode ScorchedEarth specific data
+        SEData memory fromData = appData(_fromPart.appData);
+        SEData memory toData = appData(_toPart.appData);
+
+        requireInternalCoherence(fromData);
+        requireInternalCoherence(toData);
+        requireCoreParametersUnchanged(fromData, toData);
+        requirePhaseToggle(fromData, toData);
+        requireProperAllocations(fromAllocation, toAllocation, toData);
+
         return true;
     }
+
+    function extractAllocation(VariablePart memory _variablePart)
+        private
+        pure
+        returns (Outcome.AllocationItem[] memory)
+    {
+        Outcome.OutcomeItem[] memory outcome = abi.decode(_variablePart.outcome, (Outcome.OutcomeItem[]));
+        require(outcome.length == 1, 'ScorchedEarth: Only one asset allowed');
+
+        Outcome.AssetOutcome memory assetOutcome = abi.decode(
+            outcome[0].assetOutcomeBytes,
+            (Outcome.AssetOutcome)
+        );
+
+        require(
+            assetOutcome.assetOutcomeType == uint8(Outcome.AssetOutcomeType.Allocation),
+            'ScorchedEarth: AssetOutcomeType must be Allocation'
+        );
+
+        Outcome.AllocationItem[] memory allocation = abi.decode(
+            assetOutcome.allocationOrGuaranteeBytes,
+            (Outcome.AllocationItem[])
+        );
+
+        require(
+            allocation.length == 3,
+            'ScorchedEarth: Allocation length must be 3 (Suggester, Sender, Burner)'
+        );
```

Comment:
To be clear: I assume `nParticipants=2`, despite there being 3 destinations in the outcome? It might be worth `requiring` that. 

Have you thought about the implications of this? I may have misunderstood, but if 2 out of 3 beneficiaries are in the channel and the remaining one is not, those two could collude to conclude the channel with whatever outcome they wished (and the rules in this contract can't prevent that, since[ no app logic is evaluated during a conclude](https://docs.statechannels.org/docs/wallet-dev-tutorial/execute-state-transitions#conform-to-an-on-chain-validtransition-function)). Would that be a problem, do you think?

Reply:
> To be clear: I assume nParticipants=2, despite there being 3 destinations in the outcome?

Correct. The third address is the beneficiary of any funds that are "burned" in the course of the User & Suggester's interactions.

> It might be worth requiring that.

I considered this, but wasn't sure if it was critical. As long as the clients refuse to sign a state with more than 2, is there a potential for abuse here I'm not considering? Might be worth it regardless, as gas would be minimal.

> if 2 out of 3 beneficiaries are in the channel and the remaining one is not, those two could collude to conclude the channel with whatever outcome they wished, say to cut out the 3rd beneficiary

Yes, this is definitely the case, and we are ok with it. Note the last line in [this section](https://github.com/ScopeLift/scorched-earth/blob/master/SPEC.md#suggester-story) of the spec. We figured that, while the two participants can always collude to unburn their funds, the very fact we're operating under the assumption of a behavioral "resentment" assumption makes this unlikely to happen. It's also not a big deal if it does, as they're not stealing funds from the Burner (the burner's initial balance is always 0).